File: python-bracket/app/models/game.py
from sqlalchemy import Integer, SmallInteger, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
from typing import List, Optional, TYPE_CHECKING

from .base import Base

if TYPE_CHECKING:
    from .region import Region
    from .team import Team
    from .pick import Pick


class Game(Base):
    __tablename__ = "games"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    round: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    region_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("regions.id"), nullable=True)
    team1_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("teams.id"), nullable=True)
    team2_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("teams.id"), nullable=True)
    winner_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("teams.id"), nullable=True)
    point_value: Mapped[int] = mapped_column(Integer, nullable=False)

    # Relationships
    region: Mapped[Optional["Region"]] = relationship("Region", back_populates="games")
    team1: Mapped[Optional["Team"]] = relationship("Team", foreign_keys=[team1_id])
    team2: Mapped[Optional["Team"]] = relationship("Team", foreign_keys=[team2_id])
    winner: Mapped[Optional["Team"]] = relationship("Team", foreign_keys=[winner_id])
    picks: Mapped[List["Pick"]] = relationship("Pick", back_populates="game")

    @property
    def is_completed(self) -> bool:
        return self.winner_id is not None

    @property
    def teams_set(self) -> bool:
        return self.team1_id is not None and self.team2_id is not None

    def __repr__(self) -> str:
        return f"<Game(id={self.id}, round={self.round}, completed={self.is_completed})>"

File: rust-bracket/src/models/game.rs
use chrono::{DateTime, Utc};
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "games")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub round: i16,
    pub region_id: Option<i32>,
    pub team1_id: Option<i32>,
    pub team2_id: Option<i32>,
    pub winner_id: Option<i32>,
    pub point_value: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::region::Entity",
        from = "Column::RegionId",
        to = "super::region::Column::Id"
    )]
    Region,
    #[sea_orm(
        belongs_to = "super::team::Entity",
        from = "Column::Team1Id",
        to = "super::team::Column::Id"
    )]
    Team1,
    #[sea_orm(
        belongs_to = "super::team::Entity",
        from = "Column::Team2Id",
        to = "super::team::Column::Id"
    )]
    Team2,
    #[sea_orm(
        belongs_to = "super::team::Entity",
        from = "Column::WinnerId",
        to = "super::team::Column::Id"
    )]
    Winner,
    #[sea_orm(has_many = "super::pick::Entity")]
    Picks,
}

impl Related<super::region::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Region.def()
    }
}

impl Related<super::pick::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Picks.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

pub use Entity as GameEntity;

